sql: save the ID for all the temp schemas created by a session

arulajmani opened the pull request:
Previously, we would only save the temporary schema ID for the most
recently created temporary schema. This meant that `ResolveSchemaByID`
couldn't identify any other temporary schemas. This caused descirptor
not found errors when referencing schema's across database lines.

This patch adds a map from database ID to temp schema ID for all temp
schemas which replaces the solitary temp schema ID tracking on
sessiondata.

Fixes #53163

Release note (bug fix): cross database temp schemas can now be
resolved properly

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/54636)
<!-- Reviewable:end -->

arulajmani commented:
Now that we can resolve temp schemas across databases, this uncovers the issue that temporary schemas don't have any owners because they don't have a descriptor. Has this been discussed before/do folks have thoughts around this? The obvious approach would be to special case ownership of temporary schemas everywhere (eg. alter schema, hasOwnershipOnSchema etc.) and store a more robust structure on session data than the map constructed in this patch.

solongordon reviewed (changes requested):
> Reviewed 3 of 7 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54636)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @arulajmani, @lucy-zhang, and @solongordon)

___
*[pkg/sql/authorization.go, line 628 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54636#-MHqFO1cE2aBuSb-fGei:-MHqFO1cE2aBuSb-fGej:bad6j0b) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b67f6d24d3db7c974788cb5acf33b1e65c7e6b9/pkg/sql/authorization.go#L628)):*
> ```Go
> 	case catalog.SchemaTemporary:
> 		// The user only owns the temporary schema that corresponds to the
> 		// TemporarySchemaID in the sessionData.
> ```

Comment needs an update.
___
*[pkg/sql/authorization.go, line 629 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54636#-MHqFCis0AjtDLHIqALL:-MHqFCis0AjtDLHIqALM:b5vpfoh) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b67f6d24d3db7c974788cb5acf33b1e65c7e6b9/pkg/sql/authorization.go#L629)):*
> ```Go
> 		// The user only owns the temporary schema that corresponds to the
> 		// TemporarySchemaID in the sessionData.
> 		// TODO(arul): This would mean every user has ownership over the temp schema,
> ```

But a user only owns their own temp schemas, right? This seems reasonable to me, though I don't know how it lines up with Postgres behavior.
___
*[pkg/sql/drop_table.go, line 186 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54636#-MHqFvA72ztiBUQmpVG-:-MHqFvA72ztiBUQmpVG0:b9hag4n) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b67f6d24d3db7c974788cb5acf33b1e65c7e6b9/pkg/sql/drop_table.go#L186)):*
> ```Go
> 	// Hack for temporary tables. Currently, we don't track ownership for the temp
> 	// schema, so it doesn't make sense to do this check for temporary tables.
> 	// TODO(arul): Extend ownership to apply to temp schemas.
> ```

Why can't we use `HasOwnershipOnSchema` to check if the temp schema was created by the current session?
___
*[pkg/sql/sessiondata/session_data.go, line 67 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54636#-MHqHP2JBrhm8cxsitHh:-MHqHP2JBrhm8cxsitHi:brqj9mp) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b67f6d24d3db7c974788cb5acf33b1e65c7e6b9/pkg/sql/sessiondata/session_data.go#L67)):*
> ```Go
> 	SearchPath SearchPath
> 	// DbToTempSchemaMap stores the temp schema ID for every database that has
> 	// created a temporary schema. The mapping is from decspb.ID -> descpb.ID,
> ```

s/decs/desc/
___
*[pkg/sql/sessiondata/session_data.go, line 69 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54636#-MHqGjwpALYYp3XI1Tlw:-MHqGjwpALYYp3XI1Tlx:bx6xyt5) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b67f6d24d3db7c974788cb5acf33b1e65c7e6b9/pkg/sql/sessiondata/session_data.go#L69)):*
> ```Go
> 	// created a temporary schema. The mapping is from decspb.ID -> descpb.ID,
> 	// but cannot be stored as such due to package dependencies.
> 	DbToTempSchemaMap map[uint32]uint32
> ```

Nit: No need to put "map" in the name. I'd call it `DatabaseToTempSchemaID` or maybe even just `TempSchemaIDs`.


<!-- Sent from Reviewable.io -->

thoszhang reviewed (commented):
> Reviewed 7 of 7 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54636)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @arulajmani)

___
*[pkg/sql/exec_util.go, line 2074 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54636#-MHqI8vrBIFPpc4LhvwO:-MHqI8vrBIFPpc4LhvwP:b-4tr7ck) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b67f6d24d3db7c974788cb5acf33b1e65c7e6b9/pkg/sql/exec_util.go#L2074)):*
> ```Go
> }
> 
> func (m *sessionDataMutator) SaveTemporarySchemaID(dbID uint32, tempSchemaID uint32) {
> ```

I think I'd prefer `SetTemporarySchemaIDForDatabase` or something.
___
*[pkg/sql/catalog/descs/collection.go, line 1042 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54636#-MHqcNFp5dZRE1y3-lfi:-MHqcNFq19398xoMf_aB:b9fep6g) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b67f6d24d3db7c974788cb5acf33b1e65c7e6b9/pkg/sql/catalog/descs/collection.go#L1042)):*
> ```Go
> 			Kind: catalog.SchemaTemporary,
> 			ID:   schemaID,
> 			Name: tc.sessionData.SearchPath.GetTemporarySchemaName(),
> ```

Is it possible for this ID to refer to a schema that's not for the current database?
___
*[pkg/sql/sessiondata/session_data.go, line 68 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54636#-MHqK0qZDLeqh15R1C_F:-MHqK0qZDLeqh15R1C_G:by9vaef) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b67f6d24d3db7c974788cb5acf33b1e65c7e6b9/pkg/sql/sessiondata/session_data.go#L68)):*
> ```Go
> 	// DbToTempSchemaMap stores the temp schema ID for every database that has
> 	// created a temporary schema. The mapping is from decspb.ID -> descpb.ID,
> 	// but cannot be stored as such due to package dependencies.
> ```

What exactly would the dependency cycle be if we imported `descpb`? I feel like this comment was prior to getting rid of sqlbase, etc., but I don't know if that matters.
___
*[pkg/sql/sessiondata/session_data.go, line 69 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54636#-MHqe_Pu5jfrzNG-0waP:-MHqe_Pu5jfrzNG-0waQ:bseyvj4) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b67f6d24d3db7c974788cb5acf33b1e65c7e6b9/pkg/sql/sessiondata/session_data.go#L69)):*
> ```Go
> 	// created a temporary schema. The mapping is from decspb.ID -> descpb.ID,
> 	// but cannot be stored as such due to package dependencies.
> 	DbToTempSchemaMap map[uint32]uint32
> ```

I wonder if we should unexport this and put a method on `SessionData` to get schema IDs by database ID.


<!-- Sent from Reviewable.io -->

arulajmani committed:
sql: save the ID for all the temp schemas created by a session

Previously, we would only save the temporary schema ID for the most
recently created temporary schema. This meant that `ResolveSchemaByID`
couldn't identify any other temporary schemas. This caused descirptor
not found errors when referencing schema's across database lines.

This patch adds a map from database ID to temp schema ID for all temp
schemas which replaces the solitary temp schema ID tracking on
sessiondata.

Fixes #53163

Release note (bug fix): cross database temp schemas can now be
resolved properly

arulajmani reviewed (commented):
I misunderstood the fact that users cant change in during a session, so my comment above is moot. 

This still has a couple of tests failing for types because of a bug this change uncovered -- the schemaID for pre-defined types seems to be set to 0, instead of `pg_catalog`. Still looking into fixing this, but I addressed all comments so this should be RFAL. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54636#-:-MHr2WhFCK93UPdJXGuZ:bun0bpn)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @lucy-zhang, and @solongordon)

___
*[pkg/sql/authorization.go, line 628 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54636#-MHqFO1cE2aBuSb-fGei:-MHqMNGY4WXejPZ5W4Ie:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b67f6d24d3db7c974788cb5acf33b1e65c7e6b9/pkg/sql/authorization.go#L628)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Comment needs an update.
</blockquote></details>

Done.
___
*[pkg/sql/authorization.go, line 629 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54636#-MHqFCis0AjtDLHIqALL:-MHr26KNENDp2-2MA3px:bm192m6) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b67f6d24d3db7c974788cb5acf33b1e65c7e6b9/pkg/sql/authorization.go#L629)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

But a user only owns their own temp schemas, right? This seems reasonable to me, though I don't know how it lines up with Postgres behavior.
</blockquote></details>

I misunderstood the fact that users cant change mid session -- my comments don't hold anymore. 
___
*[pkg/sql/drop_table.go, line 186 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54636#-MHqFvA72ztiBUQmpVG-:-MHr2JEo5O3t_GI85GSe:b-tw6khw) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b67f6d24d3db7c974788cb5acf33b1e65c7e6b9/pkg/sql/drop_table.go#L186)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

Why can't we use `HasOwnershipOnSchema` to check if the temp schema was created by the current session?
</blockquote></details>

Not applicable, see comment above. 
___
*[pkg/sql/exec_util.go, line 2074 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54636#-MHqI8vrBIFPpc4LhvwO:-MHqwI89DX0KyVWKg_YH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b67f6d24d3db7c974788cb5acf33b1e65c7e6b9/pkg/sql/exec_util.go#L2074)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

I think I'd prefer `SetTemporarySchemaIDForDatabase` or something.
</blockquote></details>

Done.
___
*[pkg/sql/catalog/descs/collection.go, line 1042 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54636#-MHqcNFp5dZRE1y3-lfi:-MHqweem7knbe2rBA9J7:budqbnh) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b67f6d24d3db7c974788cb5acf33b1e65c7e6b9/pkg/sql/catalog/descs/collection.go#L1042)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Is it possible for this ID to refer to a schema that's not for the current database?
</blockquote></details>

Yeah, I think it can if we're dealing with fully qualified names. One example I found is dropping a table in another database, referenced by its FQN, requires an ownership check on the table's schema. This flows through here.
___
*[pkg/sql/sessiondata/session_data.go, line 67 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54636#-MHqHP2JBrhm8cxsitHh:-MHqNorA8IjQ6mFH7Jit:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b67f6d24d3db7c974788cb5acf33b1e65c7e6b9/pkg/sql/sessiondata/session_data.go#L67)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

s/decs/desc/
</blockquote></details>

Done.
___
*[pkg/sql/sessiondata/session_data.go, line 68 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54636#-MHqK0qZDLeqh15R1C_F:-MHqy90y0fUgKQ0v7uKx:bavqvz9) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b67f6d24d3db7c974788cb5acf33b1e65c7e6b9/pkg/sql/sessiondata/session_data.go#L68)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

What exactly would the dependency cycle be if we imported `descpb`? I feel like this comment was prior to getting rid of sqlbase, etc., but I don't know if that matters.
</blockquote></details>

I think the dep cycle still exists. This is what I got:

```import stack: [github.com/cockroachdb/cockroach/pkg/cmd/docgen github.com/cockroachdb/cockroach/pkg/sql/sem/builtins github.com/cockroachdb/cockroach/pkg/sql/catalog github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb github.com/cockroachdb/cockroach/pkg/sql/parser github.com/cockroachdb/cockroach/pkg/sql/sem/tree github.com/cockroachdb/cockroach/pkg/sql/sessiondata github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb]]```
___
*[pkg/sql/sessiondata/session_data.go, line 69 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54636#-MHqe_Pu5jfrzNG-0waP:-MHqzsCEDz2R9QaT4nxa:b-tbwx8n) ([raw file](https://github.com/cockroachdb/cockroach/blob/5b67f6d24d3db7c974788cb5acf33b1e65c7e6b9/pkg/sql/sessiondata/session_data.go#L69)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

I wonder if we should unexport this and put a method on `SessionData` to get schema IDs by database ID.
</blockquote></details>

Can't unexport it because `sessionDataMutator` is defined in a different package. I'll add a method on `SessionData` as per your suggestion though and use it in `collection.go` instead of accessing the map directly.


<!-- Sent from Reviewable.io -->

solongordon reviewed (commented):
> Reviewed 5 of 12 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54636)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @arulajmani, @lucy-zhang, @otan, and @solongordon)

___
*[pkg/sql/drop_cascade.go, line 109 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/54636#-MIZwDl6C_vlzvCno8aY:-MIZwDl7A1GuX6lJIot0:b-uzu4v5) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e5a610195b5e92871df50463f08bae2e25d722c/pkg/sql/drop_cascade.go#L109)):*
> ```Go
> 			// CASCADE is temporary and was created by a different session, we can't
> 			// resolve it to check for ownership --  this allows us to circumvent that
> 			// check and avoid an error.
> ```

What is the purpose of checking `canDropTable` in this case? Is there a scenario where a `DROP ... CASCADE` should fail because other sessions have created temporary tables?


<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54636)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @lucy-zhang, @otan, and @solongordon)

___
*[pkg/sql/drop_cascade.go, line 109 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/54636#-MIZwDl6C_vlzvCno8aY:-MI_uNvG2dQV-Y2x99BQ:bg3g8y3) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e5a610195b5e92871df50463f08bae2e25d722c/pkg/sql/drop_cascade.go#L109)):*
<details><summary><i>Previously, solongordon (Solon) wrote…</i></summary><blockquote>

What is the purpose of checking `canDropTable` in this case? Is there a scenario where a `DROP ... CASCADE` should fail because other sessions have created temporary tables?
</blockquote></details>

Talked offline, but summarizing here for reference later. This check here can cause a `DROP DATABASE CASCADE` to fail because another session has created a temporary obj and the user dropping the db doesn't have `DROP` privs on the temp object. This is functionally the same as we had before. 

We're going with the conservative approach over simply skipping this check for temporary tables because it's quite rare for this to have effect -- the temp obj creator would have to explicitly remove `DROP` privileges from the user who has `DROP` privileges on the database. 


<!-- Sent from Reviewable.io -->

arulajmani commented:
TFTRs everyone! 

bors r=solongordon
